x-pack/plugins/task_manager/server/kibana_discovery_service/kibana_discovery_service.ts

```diff
@@ -0,0 +1,158 @@
+  private savedObjectsRepository: ISavedObjectsRepository;
+  private logger: Logger;
+
+  constructor({ currentNode, savedObjectsRepository, logger }: DiscoveryServiceParams) {
+    this.savedObjectsRepository = savedObjectsRepository;
+    this.logger = logger;
+    this.currentNode = currentNode;
+  }
+
+  private async upsertCurrentNode({ id, lastSeen }: DiscoveryServiceUpsertParams) {
+    await this.savedObjectsRepository.update<BackgroundTaskNode>(
+      BACKGROUND_TASK_NODE_SO_NAME,
+      id,
+      {
+        id,
+        last_seen: lastSeen,
+      },
+      { upsert: { id, last_seen: lastSeen }, refresh: true }
+    );
+  }
+
+  private async deleteInactiveNodes() {
+    const { saved_objects: inactiveNodes } =
+      await this.savedObjectsRepository.find<BackgroundTaskNode>({
+        type: BACKGROUND_TASK_NODE_SO_NAME,
+        perPage: 10000,
+        page: 1,
+        filter: `${BACKGROUND_TASK_NODE_SO_NAME}.attributes.last_seen < now-5m`,
+      });
+
+    if (inactiveNodes.length > 0) {
+      const nodesToDelete = inactiveNodes.map((node) => ({
+        type: BACKGROUND_TASK_NODE_SO_NAME,
+        id: node.attributes.id,
+      }));
+      await this.savedObjectsRepository.bulkDelete(nodesToDelete, { force: true, refresh: true });
+      this.logger.info(
+        `Inactive Kibana nodes: ${nodesToDelete.map(
+          (node) => node.id
+        )}, have been successfully deleted`
+      );
+    }
+  }
+
+  private async scheduleDeleteInactiveNodes() {
+    try {
+      await this.deleteInactiveNodes();
+      if (!this.cleanupStarted) {
+        this.logger.info('Kibana Discovery Service - Cleanup - has been started');
+        this.cleanupStarted = true;
+      }
+    } catch (e) {
+      if (!this.cleanupStarted) {
+        this.logger.error(
+          `Kibana Discovery Service - Cleanup - couldn't be started and will be retried in 1m. Error: ${e.message}`
+        );
+      } else {
+        this.logger.error(`Deleting inactive nodes failed. Error: ${e.message} `);
+      }
+    } finally {
+      setTimeout(async () => await this.scheduleDeleteInactiveNodes(), cleanupInterval);
+    }
+  }
+
+  private async scheduleUpsertCurrentNode() {
+    const lastSeen = new Date().toISOString();
+    try {
+      await this.upsertCurrentNode({ id: this.currentNode, lastSeen });
+      if (!this.discoveryStarted) {
+        this.logger.info('Kibana Discovery Service has been started');
+        this.discoveryStarted = true;
+      }
+    } catch (e) {
+      if (!this.discoveryStarted) {
+        this.logger.error(
+          `Kibana Discovery Service couldn't be started and will be retried in 10s, error:${e.message}`
+        );
+      } else {
+        this.logger.error(
+          `Background Task Node couldn't be updated. id: ${this.currentNode}, last_seen: ${lastSeen}, error:${e.message}`
+        );
+      }
+    } finally {
+      setTimeout(async () => await this.scheduleUpsertCurrentNode(), discoveryInterval);
```

Comment:
**nit:** To make it a more precise interval if ever the `upsertCurrentNode` takes a few seconds to run, should we do some math on the delay? 

Example
```
const started = new Date();
await this.upsertCurrentNode({ id: this.currentNode, lastSeen });
...
setTimeout(async () => await this.scheduleUpsertCurrentNode(), discoveryInterval - (Date.now() - started));
```